File: lib/features/health/domain/sleep.dart
import 'package:harvest/core/domain/harvest_day.dart';
import 'package:harvest/features/settings/domain/daily_cycle.dart';
import 'package:meta/meta.dart';

/// One night, as I reported it the morning after.
///
/// Both ends are times I say, not times the phone guessed: this app
/// does not watch me sleep. The target is **copied in** rather than
/// looked up later, for the same reason a session copies its targets —
/// changing my bedtime in March must not rewrite February's debt.
@immutable
class SleepNight {
  const SleepNight({
    required this.uuid,
    required this.day,
    required this.fellAsleepAt,
    required this.wokeAt,
    required this.targetMinutes,
    this.restedStars,
    this.note,
  });

  final String uuid;

  /// The Harvest Day I **woke up** on. A night is filed under its
  /// morning, because that is the day it decides how I feel.
  final HarvestDay day;

  final DateTime fellAsleepAt;
  final DateTime wokeAt;

  /// What the night was supposed to be, in minutes, as of that night.
  final int targetMinutes;

  /// 1–5, or null when I could not be bothered to say — which is
  /// itself a legitimate answer at 6 AM.
  final int? restedStars;
  final String? note;

  Duration get slept => wokeAt.difference(fellAsleepAt);
  int get sleptMinutes => slept.inMinutes;

  /// Short of target, in minutes. Negative is a surplus.
  int get shortfallMinutes => targetMinutes - sleptMinutes;
}

/// What writing the night down is worth ([[Gamification]]).
///
/// Paid for logging, not for sleeping: the app has no opinion on how
/// long I slept and every reason to want me to write it down.
const int sleepXp = 15;

/// How far back the debt looks. Two weeks: long enough for a bad run to
/// show, short enough that a bad month in the spring is not still being
/// held against me in the summer.
const int debtWindowNights = 14;

/// The running balance, minute for minute.
///
/// Every night short adds its shortfall; every long night pays it
/// down. The balance never goes below zero, because a twelve-hour
/// Sunday clears what I owe and does not put me in credit — sleep is
/// not a bank account, and pretending otherwise would let one lie-in
/// excuse the week after it.
///
/// The window is **calendar** nights, ending on [upTo] (today, or the
/// latest night logged when nobody says): the last fourteen mornings,
/// whether or not I wrote them down. Nights with nothing logged count
/// for nothing at all — an unlogged night is unknown, not a failure —
/// but they do still pass. The first cut took the fourteen most recent
/// *logged* nights, so a night from March could still weigh on July
/// for someone who logs once a week ([[Audit-v2-Beta]] B-09, N-06).
int sleepDebtMinutes(
  Iterable<SleepNight> nights, {
  int window = debtWindowNights,
  HarvestDay? upTo,
}) {
  final sorted = nights.toList()..sort((a, b) => a.day.compareTo(b.day));
  if (sorted.isEmpty) return 0;
  final end = upTo ?? sorted.last.day;
  final start = end.addDays(-(window - 1));
  final considered = sorted.where(
    (night) => night.day.compareTo(start) >= 0 && night.day.compareTo(end) <= 0,
  );

  var balance = 0;
  for (final night in considered) {
    balance += night.shortfallMinutes;
    if (balance < 0) balance = 0;
  }
  return balance;
}

/// What the debt gauge shows: how much is owed, and out of how much.
///
/// The denominator is one full night's target, so the gauge reads "how
/// many nights behind am I" rather than an unbounded number of minutes.
typedef SleepDebt = ({int minutes, double nights});

SleepDebt sleepDebt(
  Iterable<SleepNight> nights, {
  required int targetMinutes,
  int window = debtWindowNights,
  HarvestDay? upTo,
}) {
  final minutes = sleepDebtMinutes(nights, window: window, upTo: upTo);
  return (
    minutes: minutes,
    nights: targetMinutes <= 0 ? 0 : minutes / targetMinutes,
  );
}

/// The average of what was actually slept, over the logged nights only.
Duration averageSleep(Iterable<SleepNight> nights) {
  final logged = nights.toList();
  if (logged.isEmpty) return Duration.zero;
  final total = logged.fold(0, (sum, night) => sum + night.sleptMinutes);
  return Duration(minutes: total ~/ logged.length);
}

/// The night I mean to have on a given weekday.
///
/// The daily cycle is the answer for most days; a weekday with its own
/// answer overrides it. Saturday is a different night from Tuesday for
/// nearly everyone, and a target that pretends otherwise is a target
/// that gets ignored.
@immutable
class SleepTargets {
  const SleepTargets({required this.cycle, this.overrides = const {}});

  final DailyCycle cycle;

  /// Weekday (`DateTime.monday`..`DateTime.sunday`) to its own night.
  final Map<int, DailyCycle> overrides;

  DailyCycle forWeekday(int weekday) => overrides[weekday] ?? cycle;

  /// The night that ends on [day] — so the bedtime is the evening
  /// *before*, and the weekday that decides it is the morning's.
  DailyCycle forMorning(HarvestDay day) => forWeekday(day.weekday);

  int targetMinutesFor(HarvestDay day) => forMorning(day).sleep.inMinutes;

  bool get hasOverrides => overrides.isNotEmpty;
}

/// `23:30` from a stored override, and back.
///
/// Kept as text because it is a setting, and a setting that is legible
/// in the archive is worth two that are not.
String encodeCycle(DailyCycle cycle) =>
    '${_hhmm(cycle.bedTime)}-${_hhmm(cycle.wakeTime)}';

DailyCycle? decodeCycle(String? value) {
  if (value == null || value.isEmpty) return null;
  final parts = value.split('-');
  if (parts.length != 2) return null;
  final bed = _parse(parts.first);
  final wake = _parse(parts.last);
  if (bed == null || wake == null) return null;
  return DailyCycle(bedTime: bed, wakeTime: wake);
}

String _hhmm((int, int) time) =>
    '${time.$1.toString().padLeft(2, '0')}:'
    '${time.$2.toString().padLeft(2, '0')}';

(int, int)? _parse(String value) {
  final bits = value.split(':');
  if (bits.length != 2) return null;
  final hour = int.tryParse(bits.first);
  final minute = int.tryParse(bits.last);
  if (hour == null || minute == null) return null;
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null;
  return (hour, minute);
}

/// When the alarm should ring for the morning of [day].
DateTime alarmFor(HarvestDay day, SleepTargets targets) {
  final wake = targets.forMorning(day).wakeTime;
  return DateTime(day.year, day.month, day.day, wake.$1, wake.$2);
}

/// How long before the target bedtime the wind-down is said.
///
/// Half an hour: long enough to finish what I am doing, short enough
/// that it is still about tonight.
const windDown = Duration(minutes: 30);

/// When to be in bed, for the night that ends on [day].
///
/// Worked back from the alarm rather than stored separately, so it
/// wraps midnight without anyone having to think about it: waking at
/// 7 AM after eight hours means half past eleven the evening before.
DateTime bedtimeFor(HarvestDay day, SleepTargets targets) =>
    alarmFor(day, targets).subtract(targets.forMorning(day).sleep);

/// When to say it, for the night that ends on [day].
DateTime windDownFor(HarvestDay day, SleepTargets targets) =>
    bedtimeFor(day, targets).subtract(windDown);

File: packages/core/src/sleep.ts
/**
 * Sleep: how long a night was, what it was meant to be, and what is
 * owed. A port of `apps/mobile/lib/features/health/domain/sleep.dart`
 * and the bits of `daily_cycle.dart` the target depends on, held to
 * `fixtures/sleep.json`.
 *
 * Business rule #3: the debt is minute for minute, it never goes below
 * zero, it looks back fourteen **calendar** nights, and an unlogged
 * night counts for nothing in either direction.
 */

import { HarvestDay } from './harvest-day.js';

/** One night, as the `sleep_sessions` row stores it. */
export interface SleepNightLike {
  /** The Harvest Day I woke up on: a night is filed under its morning. */
  readonly harvestDay: string;
  readonly fellAsleepAt: string | Date;
  readonly wokeAt: string | Date;
  /** What the night was supposed to be, copied in when it was written. */
  readonly targetMinutes: number;
}

/** How far back the debt looks. */
export const debtWindowNights = 14;

function moment(value: string | Date): Date {
  return typeof value === 'string' ? new Date(value) : value;
}

/** Dart's `Duration.inMinutes`: whole minutes, truncated toward zero. */
export function sleptMinutes(night: SleepNightLike): number {
  return Math.trunc((moment(night.wokeAt).getTime() - moment(night.fellAsleepAt).getTime()) / 60_000);
}

/** Short of target, in minutes; negative is a surplus. */
export function shortfallMinutes(night: SleepNightLike): number {
  return night.targetMinutes - sleptMinutes(night);
}

/**
 * The running balance over the window ending on [upTo] (the latest
 * night logged when nobody says). Every short night adds its shortfall,
 * every long night pays it down, and the balance never goes below zero:
 * one twelve-hour Sunday clears what is owed and does not put me in
 * credit.
 */
export function sleepDebtMinutes(
  nights: Iterable<SleepNightLike>,
  options: { window?: number; upTo?: HarvestDay | null } = {},
): number {
  const window = options.window ?? debtWindowNights;
  const sorted = [...nights].sort((a, b) => a.harvestDay.localeCompare(b.harvestDay));
  if (sorted.length === 0) return 0;
  const end = options.upTo?.key ?? sorted[sorted.length - 1]!.harvestDay;
  const start = startKey(end, window);
  let balance = 0;
  for (const night of sorted) {
    if (night.harvestDay < start || night.harvestDay > end) continue;
    balance += shortfallMinutes(night);
    if (balance < 0) balance = 0;
  }
  return balance;
}

/** The first morning of a window of [window] nights ending on [end]. */
function startKey(end: string, window: number): string {
  const day = HarvestDay.tryParse(end);
  return day ? day.addDays(-(window - 1)).key : end;
}

export interface SleepDebt {
  readonly minutes: number;
  /** The debt in nights, so the gauge reads "how far behind am I". */
  readonly nights: number;
}

export function sleepDebt(
  nights: Iterable<SleepNightLike>,
  options: { targetMinutes: number; window?: number; upTo?: HarvestDay | null },
): SleepDebt {
  const minutes = sleepDebtMinutes(nights, options);
  return { minutes, nights: options.targetMinutes <= 0 ? 0 : minutes / options.targetMinutes };
}

/** The average of what was actually slept, over the logged nights only. */
export function averageSleepMinutes(nights: Iterable<SleepNightLike>): number {
  const logged = [...nights];
  if (logged.length === 0) return 0;
  return Math.trunc(logged.reduce((sum, night) => sum + sleptMinutes(night), 0) / logged.length);
}

// -------------------------------------------------------- the target

/** A time of day as `(hour, minute)`, the way the settings store it. */
export interface Clock {
  readonly hour: number;
  readonly minute: number;
}

/** The night I mean to have: when I go to bed, and when I get up. */
export interface Cycle {
  readonly bedTime: Clock;
  readonly wakeTime: Clock;
}

/** 11 PM to 7 AM, before anyone says otherwise (`DailyCycle.fallback`). */
export const fallbackCycle: Cycle = { bedTime: { hour: 23, minute: 0 }, wakeTime: { hour: 7, minute: 0 } };

/** `SettingsRepository.parseTime`: "HH:mm", tolerant of a missing zero pad. */
export function parseClock(raw: string | null | undefined): Clock | null {
  const parts = (raw ?? '').split(':');
  if (parts.length !== 2) return null;
  const hour = Number(parts[0]);
  const minute = Number(parts[1]);
  if (!Number.isInteger(hour) || !Number.isInteger(minute)) return null;
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null;
  return { hour, minute };
}

export function formatClock(clock: Clock): string {
  return `${String(clock.hour).padStart(2, '0')}:${String(clock.minute).padStart(2, '0')}`;
}

/** `decodeCycle`: `23:30-07:00` from a stored weekday override. */
export function decodeCycle(value: string | null | undefined): Cycle | null {
  if (!value) return null;
  const parts = value.split('-');
  if (parts.length !== 2) return null;
  const bedTime = parseClock(parts[0]);
  const wakeTime = parseClock(parts[1]);
  return bedTime && wakeTime ? { bedTime, wakeTime } : null;
}

export function encodeCycle(cycle: Cycle): string {
  return `${formatClock(cycle.bedTime)}-${formatClock(cycle.wakeTime)}`;
}

const minutesInDay = 24 * 60;

function minutesOf(clock: Clock): number {
  return clock.hour * 60 + clock.minute;
}

/**
 * How long the night is, wrapping midnight. A window that starts and
 * ends on the same minute is a whole day asleep, not none
 * (`DailyCycle.sleep`).
 */
export function cycleMinutes(cycle: Cycle): number {
  const from = minutesOf(cycle.bedTime);
  const to = minutesOf(cycle.wakeTime);
  const span = to > from ? to - from : to + minutesInDay - from;
  return span === 0 ? minutesInDay : span;
}

/**
 * The target for the morning of [day]: the daily cycle, unless that
 * weekday has a night of its own (`SleepTargets.targetMinutesFor`).
 * Saturday is a different night from Tuesday for nearly everyone.
 */
export function targetMinutesFor(
  day: HarvestDay,
  cycle: Cycle = fallbackCycle,
  overrides: Readonly<Record<number, Cycle>> = {},
): number {
  return cycleMinutes(overrides[day.weekday] ?? cycle);
}
